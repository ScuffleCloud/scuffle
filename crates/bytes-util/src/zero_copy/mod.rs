//! Zero-copy types.

mod reader;
pub use reader::*;

mod serde;
pub use serde::*;
