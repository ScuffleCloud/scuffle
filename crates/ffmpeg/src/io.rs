mod input;
mod internal;
mod output;

/// A module that contains the channel implementation for io operations.
#[cfg(feature = "channel")]
pub mod channel;

pub use input::*;
pub use output::*;
