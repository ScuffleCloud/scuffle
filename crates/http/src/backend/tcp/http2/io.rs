#![allow(dead_code)]

// TODO: Use BDP logic for bandwidth limiting & keep alive logic

use std::pin::Pin;
use std::sync::Arc;
use std::task::{Context, Poll};

use bytes::Bytes;
use spin::Mutex;

use super::ResultErrorExt;
use crate::error::{ErrorConfig, ErrorScope, ErrorSeverity};
use crate::Error;

#[derive(Clone)]
/// Shared state is shared between all streams on the same connection.
pub(crate) struct Http2SharedState {
	inner: Arc<Mutex<Http2SharedStateInner>>,
}

impl Http2SharedState {
	pub(crate) fn new(ping_pong: h2::PingPong) -> Self {
		Self {
			inner: Arc::new(Mutex::new(Http2SharedStateInner { ping_pong })),
		}
	}
}

struct Http2SharedStateInner {
	ping_pong: h2::PingPong,
}

#[derive(Clone, Copy)]
enum State {
	SizeKnown(u64),
	SizeUnknown,
	BodyComplete,
	TrailerComplete,
}

pin_project_lite::pin_project! {
	pub struct RecvStream {
		#[pin]
		io: h2::RecvStream,
		#[pin]
		state: State,
		shared: Http2SharedState,
	}
}

impl RecvStream {
	pub(crate) fn new(io: h2::RecvStream, size: Option<u64>, shared: Http2SharedState) -> Self {
		Self {
			io,
			state: size.map(State::SizeKnown).unwrap_or(State::SizeUnknown),
			shared,
		}
	}
}

impl http_body::Body for RecvStream {
	type Data = Bytes;
	type Error = crate::Error;

	fn poll_frame(
		self: Pin<&mut Self>,
		cx: &mut Context<'_>,
	) -> Poll<Option<Result<http_body::Frame<Self::Data>, Self::Error>>> {
		let mut this = self.project();

		if matches!(*this.state, State::TrailerComplete) {
			return Poll::Ready(None);
		}

		if !matches!(*this.state, State::BodyComplete) {
			match this.io.poll_data(cx) {
				Poll::Pending => return Poll::Pending,
				Poll::Ready(Some(data)) => {
					let data = data.with_config(ErrorConfig {
						severity: ErrorSeverity::Unknown,
						scope: ErrorScope::Request,
						context: "http2 body data",
					})?;

					return Poll::Ready(Some(Ok(http_body::Frame::data(data))));
				}
				Poll::Ready(None) => {
					*this.state = State::BodyComplete;
				}
			}
		}

		match this.io.poll_trailers(cx) {
			Poll::Pending => Poll::Pending,
			Poll::Ready(Ok(None)) => {
				*this.state = State::TrailerComplete;
				Poll::Ready(None)
			}
			Poll::Ready(Ok(Some(trailers))) => {
				*this.state = State::TrailerComplete;
				Poll::Ready(Some(Ok(http_body::Frame::trailers(trailers))))
			}
			Poll::Ready(Err(e)) => {
				*this.state = State::BodyComplete;
				Poll::Ready(Some(Err(Error::from(e).with_config(ErrorConfig {
					severity: ErrorSeverity::Unknown,
					scope: ErrorScope::Request,
					context: "http2 body trailers",
				}))))
			}
		}
	}

	fn is_end_stream(&self) -> bool {
		self.io.is_end_stream() || matches!(self.state, State::TrailerComplete)
	}

	fn size_hint(&self) -> http_body::SizeHint {
		match self.state {
			State::SizeKnown(size) => http_body::SizeHint::with_exact(size),
			State::SizeUnknown => http_body::SizeHint::default(),
			State::BodyComplete | State::TrailerComplete => http_body::SizeHint::with_exact(0),
		}
	}
}
