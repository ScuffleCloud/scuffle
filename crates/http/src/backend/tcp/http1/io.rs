use std::pin::Pin;
use std::task::{Context, Poll};

use bytes::{Buf, BufMut, Bytes, BytesMut};
use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt, BufReader, BufWriter};
use tokio::sync::mpsc;

use super::{proto, ResultErrorExt};
use crate::error::{ErrorConfig, ErrorScope, ErrorSeverity};

#[derive(Clone, Copy)]
enum State {
	SizeKnown(u64),
	SizeUnknown,
	BodyComplete,
	TrailerComplete,
}

pin_project_lite::pin_project! {
	pub struct RecvStream {
		#[pin]
		state: State,
		#[pin]
		inner: mpsc::Receiver<http_body::Frame<Bytes>>,
	}
}

impl RecvStream {
	pub fn new(inner: mpsc::Receiver<http_body::Frame<Bytes>>, size: Option<u64>) -> Self {
		Self {
			state: size.map_or(State::SizeUnknown, State::SizeKnown),
			inner,
		}
	}
}

impl http_body::Body for RecvStream {
	type Data = Bytes;
	type Error = std::io::Error;

	fn poll_frame(
		self: Pin<&mut Self>,
		cx: &mut Context<'_>,
	) -> Poll<Option<Result<http_body::Frame<Self::Data>, Self::Error>>> {
		let mut this = self.project();

		let remaining = match *this.state {
			State::SizeUnknown => None,
			State::SizeKnown(remaining) => Some(remaining),
			State::BodyComplete => None,
			State::TrailerComplete => return Poll::Ready(None),
		};

		match this.inner.poll_recv(cx) {
			Poll::Pending => Poll::Pending,
			Poll::Ready(Some(frame)) => {
				match frame.data_ref() {
					Some(data) => {
						if let Some(remaining) = remaining {
							if remaining >= data.len() as u64 {
								*this.state = State::SizeKnown(remaining - data.len() as u64);
							} else {
								*this.state = State::BodyComplete;
							}
						}
					}
					None => {
						// If we get trailers, the body is done.
						*this.state = State::TrailerComplete;
						return Poll::Ready(Some(Ok(frame)));
					}
				}

				Poll::Ready(Some(Ok(frame)))
			}
			Poll::Ready(None) => {
				if matches!(*this.state, State::BodyComplete | State::TrailerComplete) {
					Poll::Ready(None)
				} else {
					Poll::Ready(Some(Err(std::io::Error::new(
						std::io::ErrorKind::UnexpectedEof,
						"Request too short",
					))))
				}
			}
		}
	}

	fn is_end_stream(&self) -> bool {
		matches!(self.state, State::BodyComplete | State::TrailerComplete)
	}

	fn size_hint(&self) -> http_body::SizeHint {
		match self.state {
			State::SizeKnown(remaining) => http_body::SizeHint::with_exact(remaining),
			State::SizeUnknown => http_body::SizeHint::new(),
			State::BodyComplete => http_body::SizeHint::with_exact(0),
			State::TrailerComplete => http_body::SizeHint::with_exact(0),
		}
	}
}

pub enum BodyEncoding {
	Chunked { max_headers: usize },
	ContentLength(u64),
}

pub enum BodyParser<'a, R: AsyncRead + Unpin> {
	Empty,
	Reader {
		reader: &'a mut BufReader<R>,
		body_sender: mpsc::Sender<http_body::Frame<Bytes>>,
		encoding: BodyEncoding,
	},
}

impl<'a, R: AsyncRead + Unpin> BodyParser<'a, R> {
	pub fn empty() -> Self {
		Self::Empty
	}

	pub fn reader(
		reader: &'a mut BufReader<R>,
		body_sender: mpsc::Sender<http_body::Frame<Bytes>>,
		encoding: BodyEncoding,
	) -> Self {
		Self::Reader {
			reader,
			body_sender,
			encoding,
		}
	}

	pub async fn read(self) -> Result<(), crate::Error> {
		match self {
			Self::Empty => Ok(()),
			Self::Reader {
				reader,
				body_sender,
				encoding,
			} => match encoding {
				BodyEncoding::Chunked { max_headers } => {
					// Loop for each chunk.
					loop {
						let chunk_size = httparse::parse_chunk_size(reader.buffer()).map_err(|_| ErrorConfig {
							severity: ErrorSeverity::Debug,
							scope: ErrorScope::Connection,
							context: "parse chunk size",
						})?;

						let mut chunk_size = match chunk_size {
							httparse::Status::Complete((idx, size)) => {
								reader.consume(idx);
								size
							}
							httparse::Status::Partial => {
								if reader.buffer().len() > 128 {
									return Err(ErrorConfig {
										severity: ErrorSeverity::Debug,
										scope: ErrorScope::Connection,
										context: "chunk header too large",
									}
									.into());
								}

								reader.fill_buf().await?;
								continue;
							}
						};

						// A zero-length chunk indicates the end of the body.
						if chunk_size == 0 {
							break;
						}

						while (reader.buffer().len() as u64) < chunk_size {
							if chunk_size != 0 && reader.buffer().is_empty() {
								reader.fill_buf().await?;
								if reader.buffer().is_empty() {
									return Err(ErrorConfig {
										severity: ErrorSeverity::Debug,
										scope: ErrorScope::Connection,
										context: "Request too short",
									}
									.into());
								}
							}

							let chunk = Bytes::copy_from_slice(reader.buffer());
							reader.consume(chunk.len());
							chunk_size -= chunk.len() as u64;
							body_sender.send(http_body::Frame::data(chunk)).await.ok();
						}

						if chunk_size != 0 {
							let chunk = Bytes::copy_from_slice(&reader.buffer()[..chunk_size as usize]);
							reader.consume(chunk.len());
							body_sender.send(http_body::Frame::data(chunk)).await.ok();
						}
					}

					loop {
						let Some((idx, headers)) = proto::parse_headers(reader.buffer(), max_headers)? else {
							if reader.buffer().len() > 1024 {
								return Err(ErrorConfig {
									severity: ErrorSeverity::Debug,
									scope: ErrorScope::Connection,
									context: "trailers too large",
								}
								.into());
							}

							let size = reader.buffer().len();
							reader.fill_buf().await?;
							if reader.buffer().len() == size {
								return Err(ErrorConfig {
									severity: ErrorSeverity::Debug,
									scope: ErrorScope::Connection,
									context: "Request too short",
								}
								.into());
							}

							continue;
						};

						reader.consume(idx);
						body_sender.send(http_body::Frame::trailers(headers)).await.ok();
						break;
					}

					Ok(())
				}
				BodyEncoding::ContentLength(mut remaining) => {
					while remaining > 0 {
						reader.fill_buf().await?;

						if reader.buffer().is_empty() {
							return Err(ErrorConfig {
								severity: ErrorSeverity::Debug,
								scope: ErrorScope::Connection,
								context: "Request too short",
							}
							.into());
						}

						let should_consume = reader.buffer().len().min(remaining as usize);
						let chunk = Bytes::copy_from_slice(&reader.buffer()[..should_consume]);

						remaining -= chunk.len() as u64;
						reader.consume(chunk.len());

						// I am not sure if we should ignore errors here.
						body_sender.send(http_body::Frame::data(chunk)).await.ok();
					}

					Ok(())
				}
			},
		}
	}
}

pub async fn copy_body<E: Into<crate::Error>>(
	body: impl http_body::Body<Error = E>,
	writer: &mut BufWriter<impl AsyncWrite + Unpin>,
) -> Result<(), crate::Error> {
	let mut size = body.size_hint().exact();

	let mut body = std::pin::pin!(body);

	let mut chunked_buf = BytesMut::new();

	while let Some(frame) = std::future::poll_fn(|cx| body.as_mut().poll_frame(cx)).await {
		chunked_buf.clear();

		match frame {
			Ok(mut frame) => {
				if let Some(data) = frame.data_mut() {
					match &mut size {
						Some(remaining) => {
							while *remaining > 0 && data.has_remaining() {
								let chunk = data.chunk();
								let size = chunk.len().min(*remaining as usize);
								writer.write_all(&chunk[..size]).await.with_context("body write")?;
								*remaining -= size as u64;
								data.advance(size);
							}
						}
						None => {
							let size = data.remaining();

							// TODO we should not use std::fmt here as this is very slow & does small heap
							// allocations.
							let hex_size = format!("{:x}", size as u64);

							chunked_buf.put_slice(hex_size.as_bytes());
							chunked_buf.put_slice(b"\r\n");

							writer
								.write_all(chunked_buf.as_ref())
								.await
								.with_context("body write chunk header")?;

							// We should take advantage of vectored writes here maybe?
							while data.has_remaining() {
								let chunk = data.chunk();
								writer.write_all(chunk).await.with_context("body write chunk")?;
								data.advance(chunk.len());
							}

							writer.write_all(b"\r\n").await.with_context("body write chunk trailer")?;
							writer.flush().await.with_context("body flush")?;
						}
					}
				} else if let Some(trailers) = frame.trailers_ref() {
					if size.is_some() {
						return Err(ErrorConfig {
							severity: ErrorSeverity::Debug,
							scope: ErrorScope::Connection,
							context: "Trailers specified with non-chunked encoding",
						}
						.into());
					}

					chunked_buf.put_slice(b"0\r\n");
					proto::write_headers(&mut chunked_buf, trailers);
					writer
						.write_all(chunked_buf.as_ref())
						.await
						.with_context("body write chunk trailers")?;
					return Ok(());
				}
			}
			Err(e) => {
				return Err(e.into().with_config(ErrorConfig {
					severity: ErrorSeverity::Error,
					scope: ErrorScope::Connection,
					context: "copy body",
				}));
			}
		}

		if size.is_some_and(|s| s == 0) {
			break;
		}
	}

	if size.is_none() {
		writer
			.write_all(b"0\r\n\r\n")
			.await
			.with_context("body write chunk trailers")?;
	}

	Ok(())
}
