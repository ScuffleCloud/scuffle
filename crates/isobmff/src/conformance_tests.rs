#![cfg(test)]
#![cfg_attr(all(test, coverage_nightly), coverage(off))]

use std::collections::BTreeMap;
use std::io::BufReader;
use std::path::PathBuf;

use pretty_assertions::assert_eq;
use scuffle_bytes_util::zero_copy::{Deserialize, Serialize};

use crate::{IsoSized, IsobmffFile};

#[derive(Debug, serde_derive::Deserialize)]
struct MetadataFile {
    file_metadata: BTreeMap<String, FileMetadata>,
}

#[derive(Debug, serde_derive::Deserialize)]
struct FileMetadata {
    published: bool,
}

#[test]
fn conformance_files() {
    let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../assets")
        .join("isobmff_conformance");

    let metadata_file: MetadataFile =
        serde_json::from_reader(std::fs::File::open(dir.join("files.json")).expect("failed to open metadata file"))
            .expect("failed to deserialize metadata file");

    for (file_name, _) in metadata_file
        .file_metadata
        .into_iter()
        .filter(|(n, m)| !n.ends_with(".zip") && m.published)
    {
        println!("testing {file_name}");
        let mut file = std::fs::File::open(dir.join("files").join(&file_name)).expect("failed to open file");
        let reader = scuffle_bytes_util::zero_copy::IoRead::from(BufReader::new(&mut file));

        let isobmff_file = IsobmffFile::deserialize(reader).expect("failed to deserialize file");

        let file_size = file.metadata().expect("failed to read metadata").len() as usize;
        assert!(
            isobmff_file.size() <= file_size,
            "file size mismatch: {} (serialized size) > {} (file size)",
            isobmff_file.size(),
            file_size
        );

        let mut serialized = Vec::new();
        isobmff_file.serialize(&mut serialized).expect("failed to serialize file");
        std::fs::write(
            dir.join("output").join(file_name.split('/').last().unwrap_or(&file_name)),
            &serialized,
        )
        .expect("failed to write serialized file");
        let redeserialized_file = IsobmffFile::deserialize(scuffle_bytes_util::zero_copy::Slice::from(&serialized[..]))
            .expect("failed to deserialize serialized file");
        assert_eq!(isobmff_file, redeserialized_file, "file content mismatch");
    }
}
