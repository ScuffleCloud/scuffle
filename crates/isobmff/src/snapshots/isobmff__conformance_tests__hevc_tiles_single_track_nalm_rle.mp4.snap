---
source: crates/isobmff/src/conformance_tests.rs
expression: redeserialized_file
snapshot_kind: text
---
IsobmffFile {
    ftyp: Some(
        FileTypeBox {
            major_brand: Brand::Iso4,
            minor_version: 1,
            compatible_brands: [
                Brand::Iso4,
                Brand([104, 118, 99, 49]),
            ],
        },
    ),
    etyp: [],
    otyp: [],
    pdin: None,
    moov: Some(
        MovieBox {
            mvhd: MovieHeaderBox {
                full_header: FullBoxHeader {
                    version: 0,
                    flags: U24Be(
                        0,
                    ),
                },
                creation_time: 3559143998,
                modification_time: 3559143998,
                timescale: 600,
                duration: 6000,
                rate: 1,
                volume: 1,
                reserved1: 0,
                reserved2: 0,
                matrix: [
                    65536,
                    0,
                    0,
                    0,
                    65536,
                    0,
                    0,
                    0,
                    1073741824,
                ],
                pre_defined: [
                    0,
                    0,
                    0,
                    0,
                    0,
                    0,
                ],
                next_track_id: 2,
            },
            meta: None,
            trak: [
                TrackBox {
                    tkhd: TrackHeaderBox {
                        version: 0,
                        flags: TrackHeaderBoxFlags(
                            TrackEnabled,
                        ),
                        creation_time: 3559143998,
                        modification_time: 3559143998,
                        track_id: 1,
                        reserved1: 0,
                        duration: 6000,
                        reserved2: 0,
                        layer: 0,
                        alternate_group: 0,
                        volume: 0,
                        reserved3: 0,
                        matrix: [
                            65536,
                            0,
                            0,
                            0,
                            65536,
                            0,
                            0,
                            0,
                            1073741824,
                        ],
                        width: 1920,
                        height: 1080,
                    },
                    tref: None,
                    trgr: None,
                    edts: Some(
                        EditBox {
                            elst: Some(
                                EditListBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 1,
                                    entries: [
                                        EditListBoxEntry {
                                            edit_duration: 6000,
                                            media_time: 3000,
                                            media_rate: 1,
                                        },
                                    ],
                                },
                            ),
                        },
                    ),
                    ttyp: None,
                    meta: None,
                    mdia: MediaBox {
                        mdhd: MediaHeaderBox {
                            full_header: FullBoxHeader {
                                version: 0,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            creation_time: 3559143998,
                            modification_time: 3559143998,
                            timescale: 25000,
                            duration: 250000,
                            language: Langauge(
                                [
                                    21,
                                    14,
                                    4,
                                ],
                            ),
                            pre_defined: 0,
                        },
                        hdlr: HandlerBox {
                            full_header: FullBoxHeader {
                                version: 0,
                                flags: U24Be(
                                    0,
                                ),
                            },
                            pre_defined: 0,
                            handler_type: HandlerType::Video,
                            reserved: [
                                0,
                                0,
                                0,
                            ],
                            name: Utf8String(
                                "hvc:novpsext:tiles_rle@GPAC0.6.2-DEV-rev1025-g47a7127-master",
                            ),
                        },
                        elng: None,
                        minf: MediaInformationBox {
                            vmhd: Some(
                                VideoMediaHeaderBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            1,
                                        ),
                                    },
                                    graphicsmode: 0,
                                    opcolor: [
                                        0,
                                        0,
                                        0,
                                    ],
                                },
                            ),
                            smhd: None,
                            hmhd: None,
                            sthd: None,
                            vvhd: None,
                            unknown_boxes: [],
                            dinf: DataInformationBox {
                                dref: DataReferenceBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 1,
                                    url: [
                                        DataEntryUrlBox {
                                            full_header: FullBoxHeader {
                                                version: 0,
                                                flags: U24Be(
                                                    1,
                                                ),
                                            },
                                            location: None,
                                        },
                                    ],
                                    urn: [],
                                    imda: [],
                                    snim: [],
                                    unknown_boxes: [],
                                },
                            },
                            stbl: SampleTableBox {
                                stsd: SampleDescriptionBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 1,
                                    boxes: [
                                        UnknownBox {
                                            header: BoxHeader {
                                                size: Short(
                                                    235,
                                                ),
                                                box_type: FourCc(
                                                    "hvc2",
                                                ),
                                            },
                                            data.len: 227,
                                        },
                                    ],
                                },
                                stts: TimeToSampleBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 1,
                                    entries: [
                                        TimeToSampleBoxEntry {
                                            sample_count: 250,
                                            sample_delta: 1000,
                                        },
                                    ],
                                },
                                ctts: Some(
                                    CompositionOffsetBox {
                                        full_header: FullBoxHeader {
                                            version: 0,
                                            flags: U24Be(
                                                0,
                                            ),
                                        },
                                        entry_count: 250,
                                        entries.len: 250,
                                    },
                                ),
                                cslg: None,
                                stsc: SampleToChunkBox {
                                    full_header: FullBoxHeader {
                                        version: 0,
                                        flags: U24Be(
                                            0,
                                        ),
                                    },
                                    entry_count: 3,
                                    entries.len: 3,
                                },
                                stsz: Some(
                                    SampleSizeBox {
                                        full_header: FullBoxHeader {
                                            version: 0,
                                            flags: U24Be(
                                                0,
                                            ),
                                        },
                                        sample_size: 0,
                                        sample_count: 250,
                                        entry_size.len: 250,
                                    },
                                ),
                                stz2: None,
                                stco: Some(
                                    ChunkOffsetBox {
                                        full_header: FullBoxHeader {
                                            version: 0,
                                            flags: U24Be(
                                                0,
                                            ),
                                        },
                                        entry_count: 21,
                                        chunk_offset.len: 21,
                                    },
                                ),
                                co64: None,
                                stss: Some(
                                    SyncSampleBox {
                                        full_header: FullBoxHeader {
                                            version: 0,
                                            flags: U24Be(
                                                0,
                                            ),
                                        },
                                        entry_count: 10,
                                        sample_number: [
                                            1,
                                            26,
                                            51,
                                            76,
                                            101,
                                            126,
                                            151,
                                            176,
                                            201,
                                            226,
                                        ],
                                    },
                                ),
                                stsh: None,
                                padb: None,
                                stdp: None,
                                sdtp: None,
                                sbgp: [
                                    SampleToGroupBox {
                                        full_header: FullBoxHeader {
                                            version: 1,
                                            flags: U24Be(
                                                0,
                                            ),
                                        },
                                        grouping_type: [
                                            110,
                                            97,
                                            108,
                                            109,
                                        ],
                                        grouping_type_parameter: Some(
                                            1953655142,
                                        ),
                                        entry_count: 1,
                                        entries: [
                                            SampleToGroupBoxEntry {
                                                sample_count: 250,
                                                group_description_index: 1,
                                            },
                                        ],
                                    },
                                ],
                                sgpd: [
                                    SampleGroupDescriptionBox {
                                        full_header: FullBoxHeader {
                                            version: 1,
                                            flags: U24Be(
                                                0,
                                            ),
                                        },
                                        grouping_type: [
                                            110,
                                            97,
                                            108,
                                            109,
                                        ],
                                        default_length: Some(
                                            32,
                                        ),
                                        default_group_description_index: None,
                                        entry_count: 1,
                                        entries: [
                                            Unknown {
                                                grouping_type: [
                                                    110,
                                                    97,
                                                    108,
                                                    109,
                                                ],
                                                data: [
                                                    1,
                                                    10,
                                                    1,
                                                    0,
                                                    2,
                                                    2,
                                                    0,
                                                    3,
                                                    3,
                                                    0,
                                                    4,
                                                    4,
                                                    0,
                                                    5,
                                                    5,
                                                    0,
                                                    6,
                                                    6,
                                                    0,
                                                    7,
                                                    7,
                                                    0,
                                                    8,
                                                    8,
                                                    0,
                                                    9,
                                                    9,
                                                    0,
                                                    10,
                                                    10,
                                                    0,
                                                    0,
                                                ],
                                            },
                                        ],
                                    },
                                    SampleGroupDescriptionBox {
                                        full_header: FullBoxHeader {
                                            version: 1,
                                            flags: U24Be(
                                                0,
                                            ),
                                        },
                                        grouping_type: [
                                            116,
                                            114,
                                            105,
                                            102,
                                        ],
                                        default_length: Some(
                                            11,
                                        ),
                                        default_group_description_index: None,
                                        entry_count: 9,
                                        entries: [
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    2,
                                                    168,
                                                    0,
                                                    0,
                                                    0,
                                                    0,
                                                    2,
                                                    128,
                                                    1,
                                                    64,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    3,
                                                    168,
                                                    2,
                                                    128,
                                                    0,
                                                    0,
                                                    2,
                                                    128,
                                                    1,
                                                    64,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    4,
                                                    168,
                                                    5,
                                                    0,
                                                    0,
                                                    0,
                                                    2,
                                                    128,
                                                    1,
                                                    64,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    5,
                                                    168,
                                                    0,
                                                    0,
                                                    1,
                                                    64,
                                                    2,
                                                    128,
                                                    1,
                                                    128,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    6,
                                                    168,
                                                    2,
                                                    128,
                                                    1,
                                                    64,
                                                    2,
                                                    128,
                                                    1,
                                                    128,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    7,
                                                    168,
                                                    5,
                                                    0,
                                                    1,
                                                    64,
                                                    2,
                                                    128,
                                                    1,
                                                    128,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    8,
                                                    168,
                                                    0,
                                                    0,
                                                    2,
                                                    192,
                                                    2,
                                                    128,
                                                    1,
                                                    128,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    9,
                                                    168,
                                                    2,
                                                    128,
                                                    2,
                                                    192,
                                                    2,
                                                    128,
                                                    1,
                                                    128,
                                                ],
                                            },
                                            Unknown {
                                                grouping_type: [
                                                    116,
                                                    114,
                                                    105,
                                                    102,
                                                ],
                                                data: [
                                                    0,
                                                    10,
                                                    168,
                                                    5,
                                                    0,
                                                    2,
                                                    192,
                                                    2,
                                                    128,
                                                    1,
                                                    128,
                                                ],
                                            },
                                        ],
                                    },
                                ],
                                subs: [],
                                saiz: [],
                                saio: [],
                                csgp: [],
                            },
                        },
                        unknown_boxes: [],
                    },
                    unknown_boxes: [],
                    udta: None,
                },
            ],
            mvex: None,
            unknown_boxes: [],
            udta: None,
        },
    ),
    moof: [],
    mdat: [
        MediaDataBox {
            data.len: 672713,
        },
    ],
    imda: [],
    meta: None,
    styp: [],
    sidx: [],
    ssix: [],
    prft: [],
    unknown_boxes: [
        UnknownBox {
            header: BoxHeader {
                size: Short(
                    75,
                ),
                box_type: FourCc(
                    "free",
                ),
            },
            data.len: 67,
        },
    ],
    mfra: None,
}
